File: scripts/benchmarks/bench_real.js
#!/usr/bin/env node
const http = require('http');

const BASE_URL = 'http://localhost:8003';
const TOKEN = 'dev-token';

function api(path, payload, method) {
  return new Promise((resolve, reject) => {
    const url = new URL(path, BASE_URL);
    const data = payload ? JSON.stringify(payload) : null;
    const options = {
      hostname: url.hostname,
      port: url.port,
      path: url.pathname,
      method: method || (payload ? 'POST' : 'GET'),
      headers: {
        'Content-Type': 'application/json',
        'X-Xavier2-Token': TOKEN,
      },
    };
    if (data) options.headers['Content-Length'] = Buffer.byteLength(data);
    const req = http.request(options, (res) => {
      let body = '';
      res.on('data', c => body += c);
      res.on('end', () => {
        try { resolve(JSON.parse(body)); }
        catch { resolve(body); }
      });
    });
    req.on('error', reject);
    if (data) req.write(data);
    req.end();
  });
}

async function waitHealth() {
  for (let i = 0; i < 60; i++) {
    try {
      const r = await api('/health', null, 'GET');
      if (r.status === 'ok') return;
    } catch {}
    await new Promise(s => setTimeout(s, 1000));
  }
  throw new Error('Xavier2 not healthy');
}

async function loadDocs(docs) {
  await api('/memory/reset', {});
  const t0 = Date.now();
  for (const d of docs) {
    await api('/memory/add', {
      path: d.path, content: d.content, metadata: d.metadata || {},
      kind: d.kind, evidence_kind: d.evidence_kind,
      namespace: d.namespace, provenance: d.provenance,
    });
  }
  return (Date.now() - t0) / docs.length;
}

async function doSearch(query, filters) {
  const p = { query, limit: 5 };
  if (filters) p.filters = filters;
  const t0 = Date.now();
  const r = await api('/memory/search', p);
  return { results: r.results || [], ms: Date.now() - t0 };
}

async function doQuery(query, filters, system3) {
  const p = { query, limit: 5, system3_mode: system3 || 'disabled' };
  if (filters) p.filters = filters;
  const t0 = Date.now();
  const r = await api('/memory/query', p);
  return { response: r.response || '', ms: Date.now() - t0 };
}

async function doAgents(query, filters) {
  const p = { query, limit: 5 };
  if (filters) p.filters = filters;
  const t0 = Date.now();
  const r = await api('/agents/run', p);
  return { response: r.response || '', ms: Date.now() - t0 };
}

async function run() {
  console.log('Waiting for Xavier2...');
  await waitHealth();
  console.log('Xavier2 is healthy.');

  const fs = require('fs');
  const dataset = JSON.parse(fs.readFileSync(
    'E:\\scripts-python\\xavier2\\scripts\\benchmarks\\datasets\\internal_swal_openclaw_memory.json', 'utf-8'));

  const docs = dataset.documents;
  const cases = dataset.cases;

  console.log(`\nLoading ${docs.length} documents...`);
  const loadMs = await loadDocs(docs);
  console.log(`  Load time: ${loadMs.toFixed(1)}ms per document`);

  const results = [];
  const sLat = [], qLat = [];
  let correct = 0;

  console.log(`\nRunning ${cases.length} benchmark cases...`);
  for (const c of cases) {
    let hit = false;
    let latency = 0;

    if (c.endpoint === 'search') {
      const { results: res, ms } = await doSearch(c.query, c.filters);
      latency = ms;
      sLat.push(ms);
      const top = res[0] ? res[0].path : null;
      hit = top === c.expected_path;
      results.push({ id: c.id, endpoint: 'search', query: c.query, expected: c.expected_path, actual: top, hit, latency_ms: +ms.toFixed(1) });
    } else if (c.endpoint === 'query') {
      const { response, ms } = await doQuery(c.query, c.filters, c.system3_mode);
      latency = ms;
      qLat.push(ms);
      hit = c.expected_substring && response.toLowerCase().includes(c.expected_substring.toLowerCase());
      results.push({ id: c.id, endpoint: 'query', query: c.query, expected: c.expected_substring, actual: response.substring(0, 200), hit, latency_ms: +ms.toFixed(1) });
    } else if (c.endpoint === 'agents_run') {
      const { response, ms } = await doAgents(c.query, c.filters);
      latency = ms;
      qLat.push(ms);
      hit = c.expected_substring && response.toLowerCase().includes(c.expected_substring.toLowerCase());
      results.push({ id: c.id, endpoint: 'agents_run', query: c.query, expected: c.expected_substring, actual: response.substring(0, 200), hit, latency_ms: +ms.toFixed(1) });
    }

    if (hit) correct++;
    console.log(`  [${hit ? 'PASS' : 'FAIL'}] ${c.id} (${latency.toFixed(1)}ms)`);
  }

  const acc = (correct / cases.length * 100).toFixed(1);
  const avgS = sLat.length ? (sLat.reduce((a, b) => a + b, 0) / sLat.length).toFixed(1) : 0;
  const avgQ = qLat.length ? (qLat.reduce((a, b) => a + b, 0) / qLat.length).toFixed(1) : 0;
  const build = await api('/build', null, 'GET');

  const summary = {
    timestamp: new Date().toISOString(),
    backend: build.memory_store.backend,
    version: build.version,
    total_cases: cases.length,
    passed: correct,
    accuracy_pct: +acc,
    avg_search_ms: +avgS,
    avg_query_ms: +avgQ,
    load_ms_per_doc: +loadMs.toFixed(1),
    search_cases: sLat.length,
    query_cases: qLat.length,
  };

  console.log(`\n${'='.repeat(50)}`);
  console.log(`RESULTS -- ${summary.backend} backend (v${summary.version})`);
  console.log(`${'='.repeat(50)}`);
  console.log(`Accuracy:   ${acc}%  (${correct}/${cases.length})`);
  console.log(`Avg search: ${avgS}ms`);
  console.log(`Avg query:  ${avgQ}ms`);
  console.log(`Load/doc:   ${loadMs.toFixed(1)}ms`);

  const outDir = 'benchmark-results/real-memory-benchmark';
  require('fs').mkdirSync(outDir, { recursive: true });
  require('fs').writeFileSync(`${outDir}/summary.json`, JSON.stringify(summary, null, 2));
  require('fs').writeFileSync(`${outDir}/records.json`, JSON.stringify(results, null, 2));
  console.log(`\nSaved: ${outDir}/summary.json and records.json`);
}

run().catch(e => { console.error(e); process.exit(1); });

File: scripts/benchmarks/bench_real.py
#!/usr/bin/env python3
import json, os, time, urllib.request, urllib.error
from datetime import datetime

BASE_URL = "http://localhost:8003"
OUTPUT_DIR = "benchmark-results/real-memory-benchmark"

def get_required_xavier_token():
    for env_var in ("XAVIER_TOKEN", "XAVIER_API_KEY", "XAVIER_TOKEN"):
        token = os.environ.get(env_var, "").strip()
        if token:
            return token
    raise RuntimeError("Missing Xavier token. Set XAVIER_TOKEN, XAVIER_API_KEY, or XAVIER_TOKEN.")

TOKEN = get_required_xavier_token()

def api(path, payload=None, method="POST"):
    url = BASE_URL + path
    data = json.dumps(payload).encode("utf-8") if payload else None
    req = urllib.request.Request(url, data=data, method=method,
        headers={"Content-Type": "application/json", "X-Xavier-Token": TOKEN})
    try:
        with urllib.request.urlopen(req, timeout=60) as r:
            return json.loads(r.read().decode("utf-8"))
    except urllib.error.HTTPError as e:
        if e.code == 405 and method == "GET":
            req = urllib.request.Request(url, method="GET",
                headers={"X-Xavier-Token": TOKEN})
            with urllib.request.urlopen(req, timeout=60) as r:
                return json.loads(r.read().decode("utf-8"))
        raise

def wait_health():
    for _ in range(60):
        try:
            with urllib.request.urlopen(BASE_URL + "/health", timeout=5) as r:
                if r.status == 200: return
        except: time.sleep(1)
    raise RuntimeError("Xavier not healthy")

def load_docs(docs):
    api("/memory/reset", {})
    t0 = time.time()
    for d in docs:
        api("/memory/add", {"path":d["path"],"content":d["content"],"metadata":d.get("metadata",{}),
            "kind":d.get("kind"),"evidence_kind":d.get("evidence_kind"),
            "namespace":d.get("namespace"),"provenance":d.get("provenance")})
    return (time.time()-t0)/len(docs)*1000

def do_search(query, filters=None, limit=5):
    p = {"query":query,"limit":limit}
    if filters: p["filters"] = filters
    t0 = time.time()
    r = api("/memory/search", p)
    return r.get("results",[]), (time.time()-t0)*1000

def do_query(query, filters=None, system3="disabled"):
    p = {"query":query,"limit":5}
    if filters: p["filters"] = filters
    p["system3_mode"] = system3
    t0 = time.time()
    r = api("/memory/query", p)
    return r.get("response",""), (time.time()-t0)*1000

def do_agents(query, filters=None):
    p = {"query":query,"limit":5}
    if filters: p["filters"] = filters
    t0 = time.time()
    r = api("/agents/run", p)
    return r.get("response",""), (time.time()-t0)*1000

print("Waiting for Xavier...")
wait_health()
print("Xavier is healthy.")

with open("E:\\scripts-python\\xavier\\scripts\\benchmarks\\datasets\\internal_swal_openclaw_memory.json", encoding="utf-8") as f:
    dataset = json.load(f)

docs = dataset["documents"]
cases = dataset["cases"]

print(f"Loading {len(docs)} docs...")
load_ms = load_docs(docs)
print(f"Load time: {load_ms:.1f}ms/doc")

results = []
s_lat, q_lat = [], []
correct = 0

for case in cases:
    cid = case["id"]
    ep = case["endpoint"]
    q = case["query"]
    filt = case.get("filters")
    hit = False

    if ep == "search":
        res, ms = do_search(q, filt)
        s_lat.append(ms)
        top = res[0].get("path") if res else None
        exp = case.get("expected_path")
        hit = top == exp
        results.append({"id":cid,"endpoint":ep,"query":q,"expected":exp,"actual":top,"hit":hit,"latency_ms":round(ms,1)})
    elif ep == "query":
        resp, ms = do_query(q, filt, case.get("system3_mode","disabled"))
        q_lat.append(ms)
        exp = case.get("expected_substring","")
        hit = exp.lower() in resp.lower()
        results.append({"id":cid,"endpoint":ep,"query":q,"expected":exp,"actual":resp[:200],"hit":hit,"latency_ms":round(ms,1)})
    elif ep == "agents_run":
        resp, ms = do_agents(q, filt)
        q_lat.append(ms)
        exp = case.get("expected_substring","")
        hit = exp.lower() in resp.lower()
        results.append({"id":cid,"endpoint":ep,"query":q,"expected":exp,"actual":resp[:200],"hit":hit,"latency_ms":round(ms,1)})

    correct += 1 if hit else 0
    print(f"  {'[PASS]' if hit else '[FAIL]'} {cid} ({results[-1]['latency_ms']:.0f}ms)")

acc = correct/len(cases)*100
avg_s = sum(s_lat)/len(s_lat) if s_lat else 0
avg_q = sum(q_lat)/len(q_lat) if q_lat else 0
build = api("/build", {}, "GET")
summary = {"timestamp":datetime.now().isoformat(),"backend":build["memory_store"]["backend"],
    "version":build["version"],"total_cases":len(cases),"passed":correct,
    "accuracy_pct":round(acc,1),"avg_search_ms":round(avg_s,1),"avg_query_ms":round(avg_q,1),
    "load_ms_per_doc":round(load_ms,1),"search_cases":len(s_lat),"query_cases":len(q_lat)}

print(f"\n{'='*50}")
print(f"RESULTS -- {summary['backend']} backend (v{summary['version']})")
print(f"{'='*50}")
print(f"Accuracy:    {acc:.1f}%  ({correct}/{len(cases)})")
print(f"Avg search:  {avg_s:.1f}ms")
print(f"Avg query:   {avg_q:.1f}ms")
print(f"Load/doc:    {load_ms:.1f}ms")

import os
os.makedirs(OUTPUT_DIR, exist_ok=True)
with open(f"{OUTPUT_DIR}/summary.json","w",encoding="utf-8") as f: json.dump(summary,f,indent=2)
with open(f"{OUTPUT_DIR}/records.json","w",encoding="utf-8") as f: json.dump(results,f,indent=2)
print(f"Saved: {OUTPUT_DIR}/summary.json and records.json")
